packages/flutter/lib/src/widgets/animated_switcher.dart

```diff
@@ -215,50 +228,75 @@ class AnimatedSwitcher extends StatefulWidget {
 }
 
 class _AnimatedSwitcherState extends State<AnimatedSwitcher> with TickerProviderStateMixin {
-  final Set<_AnimatedSwitcherChildEntry> _children = new Set<_AnimatedSwitcherChildEntry>();
+  final Set<_AnimatedSwitcherChildEntry> _previousChildren = new Set<_AnimatedSwitcherChildEntry>();
   _AnimatedSwitcherChildEntry _currentChild;
+  List<Widget> _previousChildWidgetCache = const <Widget>[];
+  int serialNumber = 0;
 
   @override
   void initState() {
     super.initState();
     _addEntry(animate: false);
   }
 
-  Widget _generateTransition(Animation<double> animation) {
-    return new KeyedSubtree(
-      key: new UniqueKey(),
-      child: widget.transitionBuilder(widget.child, animation),
-    );
-  }
-
   _AnimatedSwitcherChildEntry _newEntry({
     @required AnimationController controller,
     @required Animation<double> animation,
   }) {
     final _AnimatedSwitcherChildEntry entry = new _AnimatedSwitcherChildEntry(
       widgetChild: widget.child,
-      transition: _generateTransition(animation),
+      transition: KeyedSubtree.wrap(
+        widget.transitionBuilder(
+          widget.child,
+          animation,
+        ),
+        serialNumber++,
+      ),
       animation: animation,
       controller: controller,
     );
     animation.addStatusListener((AnimationStatus status) {
       if (status == AnimationStatus.dismissed) {
-        assert(_children.contains(entry));
         setState(() {
-          _children.remove(entry);
+          _removeExpiredChild(entry);
         });
         controller.dispose();
       }
     });
     return entry;
   }
 
+  void _removeExpiredChild(_AnimatedSwitcherChildEntry child) {
+    assert(_previousChildren.contains(child));
+    assert(_previousChildWidgetCache.contains(child.transition));
+    _previousChildren.remove(child);
+    _rebuildChildWidgetCache();
+    assert(_previousChildren.length == _previousChildWidgetCache.length);
+    assert(_previousChildren.isEmpty || _previousChildren.last.transition == _previousChildWidgetCache.last);
+  }
+
+  void _retireCurrentChild() {
+    assert(!_previousChildren.contains(_currentChild));
+    assert(!_previousChildWidgetCache.contains(_currentChild.transition));
+    _currentChild.controller.reverse();
+    _previousChildren.add(_currentChild);
+    _rebuildChildWidgetCache();
+    assert(_previousChildren.length == _previousChildWidgetCache.length);
+    assert(_previousChildren.last.transition == _previousChildWidgetCache.last);
+  }
+
+  void _rebuildChildWidgetCache() {
+    _previousChildWidgetCache = new List<Widget>.unmodifiable(
```

Comment:
we call this in various places currently, with the effect that one frame could in theory invoke it multiple times.
I recommend instead just marking the cache as dirty (by setting it to null) and building it on demand (if null) in the build method (where you use it).

Reply:
OK, good idea. Done.